cloudsql/mysql/database-sql/cloudsql.go

```diff
@@ -0,0 +1,288 @@
+
+}
+
+// showTotals renders an HTML template showing the current vote totals.
+func showTotals(w http.ResponseWriter, r *http.Request) error {
+
+	totals, err := currentTotals()
+	if err != nil {
+		return err
+	}
+	err = parsedTemplate.Execute(w, totals)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// saveVote handles POST requests and saves a vote passed as http.Request form data.
+func saveVote(w http.ResponseWriter, r *http.Request) {
+	r.ParseForm()
+	team := r.Form["team"][0]
+	// [START cloud_sql_mysql_databasesql_connection]
+	sqlInsert := "INSERT INTO votes (candidate) VALUES (?)"
+	if team == "TABS" || team == "SPACES" {
+		if _, err := db.Exec(sqlInsert, team); err != nil {
+			log.Fatalf("unable to save vote: %s", err)
+		} else {
+			fmt.Fprintf(w, "Vote successfully cast for %s!\n", team)
+		}
+	}
+	// [END cloud_sql_mysql_databasesql_connection]
+}
+
+// mustGetEnv is a helper function for getting environment variables.
+// Displays a warning if the environment variable is not set.
+func mustGetenv(k string) string {
+	v := os.Getenv(k)
+	if v == "" {
+		fmt.Printf("Warning: %s environment variable not set.\n", k)
+	}
+	return v
+}
+
+// initSocketConnectionPool initializes a Unix socket connection pool for
+// a Cloud SQL instance of MySQL.
+func initSocketConnectionPool() (*sql.DB, error) {
+	// [START cloud_sql_mysql_databasesql_create_socket]
+	var (
+		dbUser                 = mustGetenv("DB_USER")
+		dbPwd                  = mustGetenv("DB_PASS")
+		instanceConnectionName = mustGetenv("INSTANCE_CONNECTION_NAME")
+		dbName                 = mustGetenv("DB_NAME")
+	)
+
+	var dbURI string
+	dbURI = fmt.Sprintf("%s:%s@unix(/cloudsql/%s)/%s", dbUser, dbPwd, instanceConnectionName, dbName)
+
+	// dbPool is the pool of database connections.
+	dbPool, err := sql.Open("mysql", dbURI)
+	if err != nil {
+		return nil, err
+	}
+
+	// [START_EXCLUDE]
+	configureConnectionPool(dbPool)
+	// [END_EXCLUDE]
+
+	return dbPool, nil
+	// [END cloud_sql_mysql_databasesql_create_socket]
+}
+
+// initTcpConnectionPool initializes a TCP connection pool for a Cloud SQL
+// instance of MySQL.
+func initTcpConnectionPool() (*sql.DB, error) {
+	// [START cloud_sql_mysql_databasesql_create_tcp]
+	var (
+		dbUser    = mustGetenv("DB_USER")
+		dbPwd     = mustGetenv("DB_PASS")
+		dbTcpHost = mustGetenv("DB_TCP_HOST")
+		dbName    = mustGetenv("DB_NAME")
+	)
+
+	var dbURI string
+	dbURI = fmt.Sprintf("%s:%s@tcp(%s)/%s", dbUser, dbPwd, dbTcpHost, dbName)
+
+	// dbPool is the pool of database connections.
+	dbPool, err := sql.Open("mysql", dbURI)
+	if err != nil {
+		return nil, err
+	}
+
+	// [START_EXCLUDE]
+	configureConnectionPool(dbPool)
+	// [END_EXCLUDE]
+
+	return dbPool, nil
+	// [END cloud_sql_mysql_databasesql_create_tcp]
+}
+
+// configureConnectionPool sets database connection pool properties.
+// For more information, see https://golang.org/pkg/database/sql
+func configureConnectionPool(dbPool *sql.DB) {
```

Comment:
One thing I notice that's still missing is [Connection Timeout](https://cloud.google.com/sql/docs/mysql/manage-connections#timeout)- I think in Go this is accomplished with Contexts, but that might make it more difficult to show in this snippet. 

Not sure if you have a good idea of how to handle this, or if we want to punt this for now and discuss in a future update.

Reply:
For the Go MySQL driver, the connection timeout setting is specified by an optional parameter on the database connection URI as covered here: https://github.com/go-sql-driver/mysql#timeout. I think we should punt this for now, because we can't simply add this to configureConnectionPool (which is called after the connection string has been passed to sql.Open), and the approach will vary across MySQL/Postgres/SQL Server. I've made a note to address this in a future update.